src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                        // cookies based on the redirectUri.
+                        string cookieHeader = GetCookieHeader(redirectUri, state.Handler._cookieContainer);
+                        if (!string.IsNullOrEmpty(cookieHeader))
+                        {
+                            if (!Interop.WinHttp.WinHttpAddRequestHeaders(
+                                state.RequestHandle,
+                                cookieHeader,
+                                (uint)cookieHeader.Length,
+                                Interop.WinHttp.WINHTTP_ADDREQ_FLAG_ADD))
+                            {
+                                WinHttpException.ThrowExceptionUsingLastError();
+                            }
+                        }
+                    }
+
+                    state.RequestMessage.RequestUri = redirectUri;
+
+                    // For security reasons, we drop the server credential if it is a 
+                    // NetworkCredential.  But we allow credentials in a CredentialCache
+                    // since they are specifically tied to URI's.
+                    if (!(state.ServerCredentials is CredentialCache))
+                    {
+                        state.ServerCredentials = null;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Interop.WinHttp.WinHttpCloseHandle(handle);
+                    state.SavedException = ex;
+                }
+            }
+            else if (internetStatus == Interop.WinHttp.WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+            {
+                if (state.RequestMessage.RequestUri.Scheme != UriSchemeHttps)
+                {
+                    // Not SSL/TLS.
+                    return;
+                }
+
+                // Grab the channel binding token (CBT) information from the request handle and put it into
+                // the TransportContext object.
+                try
+                {
+                    state.TransportContext.SetChannelBinding(state.RequestHandle);
+                }
+                catch (Exception ex)
+                {
+                    Interop.WinHttp.WinHttpCloseHandle(handle);
+                    state.SavedException = ex;
+
+                    return;
+                }
+
+                if (state.ServerCertificateValidationCallback != null)
+                {
+                    IntPtr data = IntPtr.Zero;
+                    uint dataSize = (uint)IntPtr.Size;
+
+                    if (Interop.WinHttp.WinHttpQueryOption(
+                        state.RequestHandle,
+                        Interop.WinHttp.WINHTTP_OPTION_SERVER_CERT_CONTEXT,
+                        ref data,
+                        ref dataSize))
+                    {
+                        var serverCertificate = new X509Certificate2(data);
+                        Interop.Crypt32.CertFreeCertificateContext(data);
```

Comment:
Despite having just written `X509Certificate2..ctor(IntPtr)` for Unix platforms, I did a mild double-take on this (partly because I thought `data` would be a `byte[]`).  Duplicating/duplicated handles are areas that I usually leave a comment to make both me and future readers/editors feel a bit better about knowing what is (supposed to be) going on.

```C#
// Create a managed wrapper around the certificate handle.
// Since this results in duplicating the handle, close our version after.
var serverCertificate = new X509Certificate2(data);
Interop.Crypt32.CertFreeCertificateContext(data);
```

Also, might I recommend renaming `data` to `certHandle`?